packages/labs/context/README.md

```diff
@@ -0,0 +1,169 @@
+
+export class MyElement extends LitElement {
+  @contextRequest({context: loggerContext, subscribe: true})
+  @property({attribute: false})
+  public logger?: Logger;
+
+  private doThing() {
+    this.logger?.log('a thing was done');
+  }
+}
+```
+
+Another way we can use a context in a component is via the `ContextConsumer` controller directly:
+
+#### **`my-element.ts`**:
+
+```ts
+import {ContextConsumer, property} from '@lit-labs/context';
+import {LitElement} from 'lit';
+import {Logger, loggerContext} from './logger.js';
+
+export class MyElement extends LitElement {
+  @property({attribute: false})
+  public logger?: Logger;
+
+  public constructor() {
+    // add the consumer controller
+    new ContextConsumer(
+      this,
+      loggerContext,
+      // pass a callback to update our property when we receive an update
+      (value) => {
+        this.logger = value;
+      },
+      true // pass true to get updates if the logger changes
+    );
+  }
+
+  private doThing() {
+    this.logger?.log('a thing was done');
+  }
+}
+```
+
+### Providing a Context
+
+Finally we want to be able to provide this context from somewhere higher in the DOM.
+
+Here we're using a `@contextProvider` property decorator to make a `ContextProvider`
+controller and update its value when the property value changes.
+
+#### **`my-app.ts`**:
+
+```ts
+import {LitElement} from 'lit';
+import {contextProvider} from '@lit-labs/context';
+import {loggerContext, Logger} from './my-logger.js';
+
+export class MyApp extends LitElement {
+  @contextProvider({context: loggerContext})
+  @property({attribute: false})
+  public logger: Logger = {
+    log: (msg) => {
+      console.log(`[my-app] ${msg}`);
+    },
+  });
+
+  protected render(): TemplateResult {
+    return html`<my-thing></my-thing>`;
+  }
+}
+```
+
+We can also use the `ContextProvider` controller directly:
+
+#### **`my-app.ts`**:
+
+```ts
+import {LitElement} from 'lit';
+import {ContextProvider} from '@lit-labs/context';
+import {loggerContext, Logger} from './my-logger.js';
+
+export class MyApp extends LitElement {
+  // create a provider controller and a default logger
+  private provider = new ContextProvider(this, loggerContext, {
+    log: (msg) => {
+      console.log(`[my-app] ${msg}`);
+    },
+  });
+
+  protected render(): TemplateResult {
+    return html`<my-thing></my-thing>`;
+  }
+
+  public setLogger(newLogger: Logger) {
+    // update the provider with a new logger value
+    this.provider.setValue(newLogger);
+  }
+}
+```
+
+## Known Issues
+
+### Late upgraded Context Providers
+
+In some cases you might have a context providing element that is upgraded late. LightDOM content below this provider may end up requesting a context that is currently not provided by any provider.
+
+To solve this case we provide a `ContextRoot` class which can intercept and track unsatisfied `context-request` events and then redispatch these requests when providers are updated.
```

Comment:
There's another problem here

```html
<early-provider>
  <late-provider>
    <eager-consumer>
  </late-provider>
</early-provider>
```

Oops! `<eager-consumer>` is stuck with `<early-provider>`, even though by rights it should get its own from `<late-provider>`

Does `ContextRoot` come equipped to deal with this?

My solution:

```ts
// consumer
hostConnected() {
  const event = new ContextEvent(this.context, this.contextCallback, true);
  this.host.dispatchEvent(event);
  contextEvents.set(this.host, event);
}

hostDisconnected() {
  this.dispose?.();
  this.dispose = undefined;
  contextEvents.delete(this.host);
}
```
```ts
//provider
/**
 * When a context provider connects, it listens for context-request events
 * it also fires all previously fired context-request events from their hosts,
 * in case this context provider upgraded after and is closer to a given consumer.
 */
hostConnected() {
  this.host.addEventListener('context-request', this.onChildContextEvent);
  for (const [host, fired] of contextEvents) {
    host.dispatchEvent(fired);
  }
}
```

Reply:
Have filed https://github.com/lit/lit/issues/2763 to track possibly investigating this. Not sure its actually solveable in a way that doesn't have performance impacts though and seems like a very very edge case to me?